test/core.jl

```diff
@@ -2404,6 +2404,20 @@ for f in (:(Core.arrayref), :((::typeof(Core.arrayref))), :((::Core.IntrinsicFun
     @test_throws ErrorException("cannot add methods to a builtin function") @eval $f() = 1
 end
 
+# issue #33370
+abstract type B33370 end
+
+let n = gensym(), c(x) = B33370[x][1]()
```

Comment:
For some reason I found the intent of this typed array very confusing at first sight. I think I would have been less confused by `c(x) = Ref{B33370}(x)[]()`. Especially with a comment saying that this is to prevent type inference from determining the concrete type of `x` and to force a runtime method table lookup. (at least I assume this is roughly the idea.)

Reply:
Yes, the idea is to force inference to use the abstract type in its method lookup. At some point we should change all such things in our tests to something like `infer_as(x, T)`.